### PR TITLE
Configurable metrics port names

### LINKED ISSUE #1798: Better names for metric ports of SeldonDeployment containers
SeldonDeployment defines the metrics port as `metrics` in the pod template. There are 2 such ports, one at the seldon-container-engine container and another at the predictive unit's container (for custom metrics). Since the deployment exposes multiple ports (with unknown port numbers until runtime), it would be convenient to have better names for the metrics ports. This would specifically allow defining prometheus jobs using regex to scrape the desired ports.

Ideally we'd like to have (optionally configurable) metrics port names that can be associated with seldon-specific jobs (e.g. metrics-seldon).

### PR DESCRIPTION
**Pull request by xaniasd**
Fixes #1798 and addresses #1799 (option multiple metrics ports)

Metrics ports names can be configured with environment variables and through the helm chart.

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **axsaucedo**

If they are not already assigned, you can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Hi @xaniasd. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/SeldonIO/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by axsaucedo** [comment]
Thank you for the PR @xaniasd ! 

/cc @axsaucedo @cliveseldon 
/ok-to-test
/test all

**Comment by seldondev** [comment]
Thu May  7 10:29:00 UTC 2020
The logs for [integration] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=3

**Comment by seldondev** [comment]
Thu May  7 10:29:01 UTC 2020
The logs for [notebooks] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=2

**Comment by seldondev** [comment]
Thu May  7 10:29:45 UTC 2020
The logs for [lint] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=4

**Comment by seldondev** [comment]
Thu May  7 10:30:14 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=1

**Comment by seldondev** [comment]
@xaniasd: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@xaniasd: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test integration`
notebooks | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@xaniasd: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test integration`
notebooks | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test notebooks`
pr-build | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Thu May  7 11:43:59 UTC 2020
The logs for [pr-build] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=5

**Comment by seldondev** [comment]
Thu May  7 11:44:29 UTC 2020
The logs for [lint] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=6

File: operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook_test.go [hunk 1]
```diff
@@ -394,6 +391,41 @@ func TestMetricsPortAddedTwoComponentSpecsTwoContainers(t *testing.T) {
 	g.Expect(volFound).To(BeTrue())
 }
 
+func TestOverrideMetricsPortName(t *testing.T) {
+	os.Setenv(ENV_PREDICTIVE_UNIT_METRICS_PORT_NAME, "myMetricsPort")
```
Comment (axsaucedo) [inline operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook_test.go:395]: We actually set them up as global variables so we can inject them in the test, you can do it similar to how we do it in the other tests

**Comment by seldondev** [comment]
Thu May  7 15:02:51 UTC 2020
The logs for [integration] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=7

File: operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go [hunk 2]
```diff
@@ -44,6 +45,7 @@ var (
 	C                                   client.Client
 	envPredictiveUnitServicePort        = os.Getenv(ENV_PREDICTIVE_UNIT_SERVICE_PORT)
 	envPredictiveUnitServicePortMetrics = os.Getenv(ENV_PREDICTIVE_UNIT_SERVICE_PORT_METRICS)
+	predictiveUnitMetricsPortName       = GetEnv(ENV_PREDICTIVE_UNIT_METRICS_ENDPOINT, constants.DefaultMetricsPortName)
```
[the remark this answers is not in the document]
Comment (xaniasd) [reply operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go]: will do

File: operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook_test.go [hunk 3]
```diff
@@ -208,10 +210,9 @@ func TestDefaultSingleContainer(t *testing.T) {
 	spec.DefaultSeldonDeployment("mydep", "default")
 
 	// Test Metric Ports
-	metricPort := GetPort(constants.MetricsPortName, spec.Predictors[0].ComponentSpecs[0].Spec.Containers[0].Ports)
+	metricPort := GetPort(constants.DefaultMetricsPortName, spec.Predictors[0].ComponentSpecs[0].Spec.Containers[0].Ports)
```
[the remark this answers is not in the document]
Comment (xaniasd) [reply operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook_test.go:213]: will do

File: operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook_test.go (hunk 1 above)
Comment (xaniasd) [reply operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook_test.go:395]: I didn't find a test overriding env vars, can you point me to one?

**Comment by seldondev** [comment]
Thu May  7 17:16:31 UTC 2020
The logs for [lint] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=9

**Comment by seldondev** [comment]
Thu May  7 17:16:44 UTC 2020
The logs for [pr-build] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=8

**Comment by seldondev** [comment]
@xaniasd: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test notebooks`
pr-build | 496c039e5be553182ac058fb1d684c9a6a4bbc27 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Thu May  7 17:25:36 UTC 2020
The logs for [pr-build] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=10

**Comment by seldondev** [comment]
@xaniasd: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test notebooks`
pr-build | 496c039e5be553182ac058fb1d684c9a6a4bbc27 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Thu May  7 17:58:46 UTC 2020
The logs for [lint] [12] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/12.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=12

**Comment by seldondev** [comment]
Thu May  7 17:58:47 UTC 2020
The logs for [pr-build] [11] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/11.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=11

**Comment by seldondev** [comment]
@xaniasd: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test notebooks`
pr-build | 309d3e1cbbe63b0b2c2a00cd0b04241832c5e50e | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Thu May  7 18:24:12 UTC 2020
The logs for [pr-build] [13] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/13.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=13

**Comment by seldondev** [comment]
@xaniasd: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test notebooks`
pr-build | 309d3e1cbbe63b0b2c2a00cd0b04241832c5e50e | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by xaniasd** [comment]
hmm, I think I'll need some support on this. The only failing step I can see is in the following part:

```
___________________ test_model_template_app_grpc_tags[False] ___________________
Traceback (most recent call last):
  File "/workspace/source/python/tests/test_microservice.py", line 221, in test_model_template_app_grpc_tags
    response = stub.Predict(request=request)
  File "/workspace/source/python/.tox/py/lib/python3.7/site-packages/grpc/_channel.py", line 826, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/workspace/source/python/.tox/py/lib/python3.7/site-packages/grpc/_channel.py", line 729, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Socket closed"
	debug_error_string = "{"created":"@1588874535.253259506","description":"Error received from peer ipv4:0.0.0.0:5000","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Socket closed","grpc_status":14}"
``` 

this has been failing since my last 2 commits, in which only unit tests are touched.

**Comment by cliveseldon** [comment]
> hmm, I think I'll need some support on this. The only failing step I can see is in the following part:
> 
> ```
> ___________________ test_model_template_app_grpc_tags[False] ___________________
> Traceback (most recent call last):
>   File "/workspace/source/python/tests/test_microservice.py", line 221, in test_model_template_app_grpc_tags
>     response = stub.Predict(request=request)
>   File "/workspace/source/python/.tox/py/lib/python3.7/site-packages/grpc/_channel.py", line 826, in __call__
>     return _end_unary_response_blocking(state, call, False, None)
>   File "/workspace/source/python/.tox/py/lib/python3.7/site-packages/grpc/_channel.py", line 729, in _end_unary_response_blocking
>     raise _InactiveRpcError(state)
> grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
> 	status = StatusCode.UNAVAILABLE
> 	details = "Socket closed"
> 	debug_error_string = "{"created":"@1588874535.253259506","description":"Error received from peer ipv4:0.0.0.0:5000","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Socket closed","grpc_status":14}"
> ```
> 
> this has been failing since my last 2 commits, in which only unit tests are touched.

Hmm ok. We have seen flakiness on another test: https://github.com/SeldonIO/seldon-core/issues/1745 I wonder if its similar @RafalSkolasinski ?

**Comment by axsaucedo** [comment]
Yes that test is known to be flaky - @cliveseldon linked the issue where we're addressing it in the next release. For now we'll re-run (as the previous pr-build passed).
/test this
/test integration

**Comment by seldondev** [comment]
Fri May  8 06:50:46 UTC 2020
The logs for [integration] [15] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/15.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=15

**Comment by axsaucedo** [comment]
Thanks for addressing the comments btw - looks good!

**Comment by seldondev** [comment]
Fri May  8 06:52:09 UTC 2020
The logs for [pr-build] [14] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/14.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=14

**Comment by seldondev** [comment]
@xaniasd: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 031c823f34ab224d637a90156c241f6f5ca56041 | [link]() | `/test notebooks`
pr-build | 309d3e1cbbe63b0b2c2a00cd0b04241832c5e50e | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Fri May  8 15:26:41 UTC 2020
The logs for [pr-build] [16] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/16.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=16

**Comment by xaniasd** [comment]
seems to have passed now @axsaucedo!

**Comment by axsaucedo** [comment]
You're right @xaniasd thank you very much for your contribution!
/approve

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/1806#issuecomment-627954238" title="Approved">axsaucedo</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Wed May 13 12:39:38 UTC 2020
The logs for [lint] [18] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/18.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=18

**Comment by seldondev** [comment]
Wed May 13 12:39:42 UTC 2020
The logs for [pr-build] [17] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1806/17.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1806 --build=17

**Comment by seldondev** [comment]
failed to trigger Pull Request pipeline
* failed to create agent
* failed to calculate in repo config
* failed to load trigger config for repository SeldonIO/seldon-core for ref 309d3e1cbbe63b0b2c2a00cd0b04241832c5e50e
* failed to find any lighthouse configuration files in repo SeldonIO/seldon-core at sha 309d3e1cbbe63b0b2c2a00cd0b04241832c5e50e
* failed to process repo SeldonIO/seldon-core refref 309d3e1cbbe63b0b2c2a00cd0b04241832c5e50e
* failed to list files in directory /var/tmp/gitrepo660600678/.lighthouse
* open /var/tmp/gitrepo660600678/.lighthouse
* no such file or directory

**Pull request merged**